No longer require a `Procfile` for Godeps and Govendor

RochesterinNYC opened the pull request:
- This makes the go buildpack run the generated main package binary name to start the web worker if no  procfile is present and no start command is specified

freeformz commented on bin/compile:
```diff
@@ -384,6 +384,11 @@ case $TOOL in
     ;;
 esac
 
+if [ ! -z ${name+x} ];
```
When will this ever be false?

RochesterinNYC replied:
In the case of the gb package manager being used, it seems that `name` would not be set.

freeformz commented on bin/compile:
```diff
@@ -384,6 +384,11 @@ case $TOOL in
 
+if [ ! -z ${name+x} ];
+then
+  echo -e "---\ndefault_process_types:\n  web: $(basename $name)" >> /tmp/buildpack-release-step.yml
```
This assumes $(basename $name) is the executable. This only works when $name is also the main package. I find this is rarely the case in anything except smaller apps.

RochesterinNYC replied:
The assumption that I was making is that a user would not be pushing an app without a `Procfile` unless that app is a simple/smaller app that would result in one binary (from the main package) being generated, with the intention that this generated binary would be run for the app. If the app was more complex and would result in multiple binaries being generated, then the user would include a `Procfile` to start different worker types off of the different binaries. Are there inaccuracies with this assumption?

freeformz commented:
Thanks for your interest!

Please see my comments.

Perhaps it would be better to look into bin and if there is only a single executable, use that instead, assuming there is no Procfile.

freeformz commented:
Also from our docs:

`Of course, rather than using default_process_types, it’s simpler to just write a default Procfile if one isn’t provided.`

I'd rather do that TBH.

RochesterinNYC commented:
Okay, I agree with the generating a default `Procfile` rather than changing `bin/release` point. I'll update the PR to reflect that.

freeformz commented on bin/compile:
```diff
@@ -384,6 +384,11 @@ case $TOOL in
     ;;
 esac
 
+if ! test -e $build/Procfile && ! test -z ${name+x}
```
`test -n "${name}"` instead of `! test -z ${name+x}` ?

RochesterinNYC committed:
No longer require a `Procfile` for Godeps and Govendor

- Run the generated main package binary name if no procfile present and
  no start command is specified

RochesterinNYC committed:
Add unit test for default Procfile generation for Godeps and Govendor

RochesterinNYC commented:
Updated this PR, addressed the comments, and added unit tests.

freeformz commented:
I am the primary maintainer and I was out on vacation. I'll look at this more soon.

freeformz commented:
:+1:

freeformz commented:
Thanks for the contribution!
